src/framework/draw/internal/qpainterprovider.cpp

```diff
@@ -357,6 +357,26 @@ void QPainterProvider::setClipRect(const RectF& rect)
     m_painter->setClipRect(rect.toQRectF());
 }
 
+void QPainterProvider::setMask(const mu::engraving::Shape& shape)
+{
+    if (shape.empty()) {
+        m_painter->setClipPath(QPainterPath(), Qt::NoClip);
+        return;
+    }
+
+    QPainterPath background;
+    background.addRect(QRectF(-1000000.0, -1000000.0, 2000000.0, 2000000.0)); // Start from arbitrary large area
+
+    QPainterPath exclusionRegion;
+    for (const RectF& rect : shape.elements()) {
+        exclusionRegion.addRect(rect.toQRectF());
+    }
+
+    QPainterPath mask = background.subtracted(exclusionRegion);
+
+    m_painter->setClipPath(mask);
```

Comment:
Do you know how this affects performance? I think I heard at some point that clipping is relatively expensive, and I wonder if the size of the clipping area contributes to that or not. Perhaps there's no noticeable difference in our case, but it would be good to do a bit of testing, for example by looking at how smoothly the score moves when scrolling/dragging/zooming quickly.

Reply:
Rectangular clipping like this is ["guaranteed" to be efficient](https://doc.qt.io/qt-6/qpainter.html#clipping:~:text=Rectangular%20clipping%20with%20simple%20transformations%20and%20intersect%20clip.). Also, intuition tells me that the size of the clipping area may only affect performance for rasterized graphics, not vector drawing.